### PR TITLE
chore: Removing tests/requirements.txt and adding dependencies to the tox.ini file

### PR DESCRIPTION
**Pull request by lucas-belo**
Resolves https://github.com/scrapy/scrapy/issues/6270

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -9,7 +9,23 @@ minversion = 1.7.0
...
+    testfixtures
+    uvloop; platform_system != "Windows"
+
+    bpython  # optional for shell wrapper tests
+    brotli; implementation_name != 'pypy'  # optional for HTTP compress downloader middleware tests
+    # 1.1.0 is broken on PyPy: https://github.com/google/brotli/issues/1072
+    brotli==1.0.9; implementation_name == 'pypy'  # optional for HTTP compress downloader middleware tests
+    zstandard; implementation_name != 'pypy'  # optional for HTTP compress downloader middleware tests
+    ipython
+    pywin32; sys_platform == "win32"
```
Comment (Gallaecio) [inline tox.ini:28]: These need to be moved to extra-deps: `brotlicffi` (added after you created your PR) , `zstandard`, `uvloop`, `bpython`, `ipython`.

These are required by some non-skippable tests and need to stay here: `pytest` and its plugins, `sybil`, [`pexpect`](https://github.com/scrapy/scrapy/commit/d19e315b0b07b070f45cf3c082d4dfe0e6a87186), `mitmproxy`, `testfixtures`, [`pyftpdlib`](https://github.com/scrapy/scrapy/commit/e70975f0bb4e8378bf0e00ae7a7014247cd59441), [`pywin32`](https://github.com/scrapy/scrapy/commit/19ad94105f70a26ce47683d488b66fb77b79067b).

Comment (lucas-belo) [reply tox.ini:28]: Doing it!

Comment (wRAR) [reply tox.ini:28]: I made pyftpdlib optional temporarily when it wasn't ready for 3.12, we can reintroduce that code if we want. mitmproxy tests are also skippable I think?

Comment (wRAR) [reply tox.ini:28]: As an aside, I wondered before if we can get rid of `testfixtures` and use pytest's `caplog` instead, but I'm not sure how to use it as I think we don't have fixture injections in our `unittest` tests.

**Comment by Gallaecio** [comment]
Note that you’ll need to [resolve conflicts](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github).

**Comment by lucas-belo** [comment]
Trying to find the lowest version of the packages with which tests pass, with a binary search strategy among all the packages versions.

File: tox.ini [hunk 2]
```diff
@@ -130,6 +146,11 @@ deps =
 [testenv:extra-deps-pinned]
 basepython = python3.8
 deps =
+    uvloop==0.19.0
+    zstandard==0.22.0
+    bpython==0.24
+    ipython==8.10.0
```
Comment (Gallaecio) [inline tox.ini:152]: These seem to be the latest versions. They should instead be the minimum versions for which tests pass.

File: tox.ini [hunk 3]
```diff
@@ -9,7 +9,18 @@ minversion = 1.7.0
...
+    pexpect >= 4.8.0
+    pyftpdlib >= 1.5.8
+    pytest
+    pytest-cov==4.0.0
+    pytest-xdist
+    uvloop; platform_system != "Windows"
+    sybil >= 1.3.0  # https://github.com/cjw296/sybil/issues/20#issuecomment-605433422
+    testfixtures
+    brotli; implementation_name != 'pypy'  # optional for HTTP compress downloader middleware tests
+    brotlicffi; implementation_name == 'pypy'  # optional for HTTP compress downloader middleware tests
```
Comment (Gallaecio) [inline tox.ini:22]: uvloop, brotli and brotlicffi should be moved into `extra-deps`.

File: tox.ini (hunk 2 above)
Comment (lucas-belo) [reply tox.ini:152]: I know.

I was just making tests, because when I did like you said, the CI tests was not approved

File: tox.ini (hunk 3 above)
Comment (lucas-belo) [reply tox.ini:22]: The same: I was just making tests, because when I did like you said, the CI tests was not approved

**Comment by Gallaecio** [comment]
Waiting for CI, given we have to manually approve every time, might be too time-consuming. You can run tests locally with `tox`, e.g. `tox -e py` for regular tests with your system version of Python, `tox -e extra-deps-pinned` and `tox -e extra-deps` for the extra-deps environments, etc.

File: tox.ini (hunk 1 above)
Comment (lucas-belo) [reply tox.ini]: Hey @Gallaecio, I was making some tests and I see that:

Removing `bpython` from the [testenv]deps, causes fail on the following tests:

`FAILED tests/test_commands.py::RunSpiderCommandTest::test_custom_asyncio_loop_enabled_true
FAILED tests/test_crawler.py::CrawlerProcessSubprocess::test_asyncio_enabled_reactor_different_loop
FAILED tests/test_crawler.py::CrawlerProcessSubprocess::test_asyncio_enabled_reactor_same_loop
FAILED tests/test_crawler.py::CrawlerProcessSubprocess::test_custom_loop_asyncio
FAILED tests/test_crawler.py::CrawlerProcessSubprocess::test_custom_loop_asyncio_deferred_signal`

and these:

`FAILED tests/test_utils_display.py::TestDisplay::test_pformat - AssertionError: "{'a': 1}" not found in {"{\x1b[33m'\x1b[39;49;00m\x1b[33ma\x1b[39;49;00m\x1b[33m'\x1b[39;49;00m: \x1b[34m1\x1b[39;49;00m}\x1b[37m\x1b[39;49;00m\n",...
FAILED tests/test_utils_display.py::TestDisplay::test_pformat_old_windows - AssertionError: "{'a': 1}" not found in {"{\x1b[33m'\x1b[39;49;00m\x1b[33ma\x1b[39;49;00m\x1b[33m'\x1b[39;49;00m: \x1b[34m1\x1b[39;49;00m}\x1b[37m\x1b[39;49;00m\n",...
FAILED tests/test_utils_display.py::TestDisplay::test_pformat_windows - AssertionError: "{'a': 1}" not found in {"{\x1b[33m'\x1b[39;49;00m\x1b[33ma\x1b[39;49;00m\x1b[33m'\x1b[39;49;00m: \x1b[34m1\x1b[39;49;00m}\x1b[37m\x1b[39;49;00m\n",...
`

Can we keep bpython on the [testenv]deps?

Comment (lucas-belo) [reply tox.ini]: Also, I still have to change the versions of the [extra-deps-pinned]deps to the lowest possible version, I was having other troubles because of this problems with bpython and uvloop, I am resuming finding this versions.

Comment (lucas-belo) [reply tox.ini]: I have already pinned the lowest version for the libs:

zstandard==0.1
ipython==2.0.0
brotli==0.5.2
brotlicffi==0.8.0

**Comment by lucas-belo** [comment]
Hey @Gallaecio and @wRAR, all tests were successful for the last commit on my fork's CI, is there any difference? Maybe re-run the failed tests?

https://github.com/lucas-belo/scrapy/actions/runs/8211648105
image

**Comment by wRAR** [comment]
3.8 tests fail because of #6286

File: tox.ini (hunk 1 above)
Comment (Gallaecio) [reply tox.ini]: Regarding tests not passing when `uvloop` or `bpython` are missing, we should mark those tests to be skipped when those packages are not installed.

For `uploop`, that means editing [this](https://github.com/scrapy/scrapy/blob/642af407049a5ce8e76b7999c68333670c8c8622/conftest.py#L84-L95) and replacing its implementation with something like:

```
@pytest.fixture(autouse=True)
def requires_uvloop(request):
    if not request.node.get_closest_marker("requires_uvloop"):
        return
    try:
        import uvloop
    except ImportError:
        pytest.skip("uvloop is not installed")
```

For `bpython`, you can see [here](https://github.com/scrapy/scrapy/blob/642af407049a5ce8e76b7999c68333670c8c8622/tests/test_utils_console.py#L30) how to mark a test to be skipped when it is missing.

File: tox.ini [hunk 4]
```diff
@@ -124,8 +141,14 @@ deps =
     robotexclusionrulesparser
     Pillow
     Twisted[http2]
-    brotli
-    zstandard
+;    brotli
+;    zstandard
```
Comment (Gallaecio) [inline tox.ini:145]: These lines can be deleted now.

File: tox.ini [hunk 5]
```diff
@@ -136,6 +159,12 @@ deps =
     Pillow==7.1.0
     robotexclusionrulesparser==1.6.2
     brotlipy
+
+
+    zstandard==0.1; implementation_name != 'pypy'
+    ipython==2.0.0
+    brotli==0.5.2; implementation_name != 'pypy'
+    brotlicffi==0.8.0; implementation_name != 'pypy'
```
Comment (Gallaecio) [inline tox.ini:167]:
```suggestion
    brotlicffi==0.8.0; implementation_name == 'pypy'
```

File: tox.ini (hunk 4 above)
Comment (lucas-belo) [reply tox.ini]: Done

File: tox.ini (hunk 1 above)
Comment (lucas-belo) [reply tox.ini]: Take a look please

Comment (Gallaecio) [reply tox.ini]: Looks good!

Comment (Gallaecio) [reply tox.ini]: > I made pyftpdlib optional temporarily when it wasn't ready for 3.12, we can reintroduce that code if we want. mitmproxy tests are also skippable I think?

@wRAR I think it is OK to treat these 2 the same as pytest, i.e. required for tests.

**Review by Gallaecio — APPROVED** [review]
Great work!

I think we can merge after a run of `pre-commit run --all` to fix the pre-commit CI job, and provided no other *unexpected* CI job failures.

File: tests/test_utils_display.py [hunk 6]
```diff
@@ -1,9 +1,18 @@
...
+try:
+    import bpython
 
+    bpy = True
+    del bpython
+except ImportError:
+    bpy = False
+
+
+@skipIf(not bpy, "bpython not available in testenv")
```
Comment (wRAR) [inline tests/test_utils_display.py:15]: Why is this needed? Looks like this test works without bpython and doesn't use it?

Comment (lucas-belo) [reply tests/test_utils_display.py:15]: Without this, we have errors on this tests:

image

Like a said on this comment: https://github.com/scrapy/scrapy/pull/6272#discussion_r1518409690

Do you think there is a better way?

**Comment by lucas-belo** [comment]
> Great work!
> 
> I think we can merge after a run of `pre-commit run --all` to fix the pre-commit CI job, and provided no other _unexpected_ CI job failures.

Done!

**Comment by Gallaecio** [comment]
I think the issues in extra-deps-pinned may be real issues. We might need a higher version of some extra dep.

**Comment by wRAR** [comment]
Of uvloop, it seems: https://github.com/MagicStack/uvloop/issues/126

**Comment by lucas-belo** [comment]
I am gonna see it

**Comment by lucas-belo** [comment]
With uvloop==0.12.0, all tests passed, but we got 4 erros:

image

**Comment by lucas-belo** [comment]
With uvloop==0.14.0 we got all tests successful and no errors, I think this is the right one:

image

**Comment by Gallaecio** [comment]
I have added pygments, and I squashed your commits because although `git diff scrapy/master` showed the right changes but the GitHib diff included 30 files (I believe because you merged a PR of mine to fix CI jobs, which I then squash-merged into the main branch), making review online a bit harder.

**Review by wRAR — APPROVED** [review]

**Comment by Gallaecio** [comment]
@lucas-belo Thank you for your amazing work here!
